rpc/conn.c

```diff
@@ -1454,8 +1454,9 @@ M0_INTERNAL void m0_rpc_conn_ha_timer_stop(struct m0_rpc_conn *conn)
 	if (m0_sm_timer_is_armed(&conn->c_ha_timer)) {
 		M0_LOG(M0_DEBUG, "Cancelling HA timer; rpc conn=%p", conn);
 		m0_sm_timer_cancel(&conn->c_ha_timer);
-	} else if (M0_IN(conn->c_ha_timer.tr_timer.t_state,
-			(M0_TIMER_STOPPED, M0_TIMER_INITED)))
+	}
+	if (M0_IN(conn->c_ha_timer.tr_timer.t_state,
+		 (M0_TIMER_STOPPED, M0_TIMER_INITED)))
```

Comment:
Is it possible this timer in other state? e.g. M0_TIMER_UNINIT?
I can not be in M0_TIMER_RUNNING, because if it is ARMED, it is cancelled now.

If it can not be in other states, we then need to assert this condition.

Reply:
@huanghua78 with this log message
```
[root@ssc-vm-g2-rhev4-1290 cortx-motr]# git diff
diff --git a/rpc/conn.c b/rpc/conn.c
index bfc85ea..fa60d38 100644
--- a/rpc/conn.c
+++ b/rpc/conn.c
@@ -1451,6 +1451,7 @@ M0_INTERNAL void m0_rpc_conn_ha_timer_stop(struct m0_rpc_conn *conn)
 {
        M0_ENTRY("conn %p", conn);
        M0_PRE(m0_rpc_machine_is_locked(conn->c_rpc_machine));
+       M0_LOG(M0_ALWAYS, "TIMER STATE :%d", conn->c_ha_timer.tr_timer.t_state);
        if (m0_sm_timer_is_armed(&conn->c_ha_timer)) {
                M0_LOG(M0_DEBUG, "Cancelling HA timer; rpc conn=%p", conn);
                m0_sm_timer_cancel(&conn->c_ha_timer);
[root@ssc-vm-g2-rhev4-1290 cortx-motr]#
```
we see that the function m0_rpc_conn_ha_timer_stop() is being called in various states of timer. Atleast in fs_stats UT, it should be the same case for normal running cluster.
```
[root@ssc-vm-g2-rhev4-1290 cortx-motr]# ./scripts/m0 run-ut -t spiel-ci-ut:stats &> fsstats_ut.log
[root@ssc-vm-g2-rhev4-1290 cortx-motr]# cat fsstats_ut.log | grep "TIMER STATE :0" | wc -l
498
[root@ssc-vm-g2-rhev4-1290 cortx-motr]# cat fsstats_ut.log | grep "TIMER STATE :1" | wc -l
0
[root@ssc-vm-g2-rhev4-1290 cortx-motr]# cat fsstats_ut.log | grep "TIMER STATE :2" | wc -l
48
[root@ssc-vm-g2-rhev4-1290 cortx-motr]# cat fsstats_ut.log | grep "TIMER STATE :3" | wc -l
12
```

Reply:
For reference
```
enum m0_timer_state {
        /** Not initialized. */
        M0_TIMER_UNINIT = 0,
        /** Initialized. */
        M0_TIMER_INITED,
        /** Timer is running. */
        M0_TIMER_RUNNING,
        /** Timer is stopped */
        M0_TIMER_STOPPED,
        /** Number of timer states */
        M0_TIMER_STATE_NR,
};
```

Reply:
If the state is M0_TIMER_RUNNING at the beginning of this function, then it will stopped in this function when m0_sm_timer_cancel() is called. So, the final state should be M0_TIMER_STOPPED.

In some cases it is in M0_TIMER_UNINIT. This might happen in UT, I think, because in some test cases the timer is not fully initialized.

OK. The checking here is necessary and suitable.